In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris

C:\Users\dell4\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:61: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [2]:
iris = load_iris()
# iris

In [3]:
df = pd.DataFrame(data= iris.data, columns=iris.feature_names)
df['target'] = iris.target
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [4]:
print('Dataset Shape:', df.shape)

Dataset Shape: (150, 5)


In [5]:
print('Columns:', df.columns.tolist())

Columns: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)', 'target']


In [6]:
print("target Calsses:", np.unique(df['target']))

target Calsses: [0 1 2]


In [7]:
print("Species Names:", iris.target_names)

Species Names: ['setosa' 'versicolor' 'virginica']


**150 rows** → means 150 flower samples

**5 columns** → 4 features (sepal & petal measurements) + 1 target (species)

And the **target (species)** has 3 classes:

0 → Setosa

1 → Versicolor

2 → Virginica

In [8]:
print(df.describe())

       sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000  


In [9]:
from sklearn.model_selection import train_test_split

In [10]:
X = df.drop('target', axis = 1)
y = df['target']

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42, stratify=y)

In [12]:
print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

Training samples: 120
Testing samples: 30


In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler

In [14]:

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [15]:
# Models
models = {
    "Logistic Regression": LogisticRegression(max_iter=200),
    "SVM": SVC(),
    "Decision Tree": DecisionTreeClassifier(random_state=42)
}

results = []
predictions = {}

In [16]:
# Train & evaluate
for name, model in models.items():
    if name in ["Logistic Regression", "SVM"]:
        model.fit(X_train_scaled, y_train)
        y_pred = model.predict(X_test_scaled)
    else:
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
    
    acc = accuracy_score(y_test, y_pred)
    results.append({"model": name, "accuracy": acc})
    predictions[name] = y_pred

In [17]:
# Show accuracy comparison
results_df = pd.DataFrame(results).sort_values("accuracy", ascending=False)
print("Model accuracies (higher is better):")
print(results_df.to_string(index=False))

Model accuracies (higher is better):
              model  accuracy
                SVM  0.966667
Logistic Regression  0.933333
      Decision Tree  0.933333


In [18]:
# Pick best model
best_name = results_df.iloc[0]["model"]
best_pred = predictions[best_name]
print(f"\nBest model: {best_name} (accuracy = {results_df.iloc[0]['accuracy']:.2f})")


Best model: SVM (accuracy = 0.97)


In [19]:
# Detailed metrics for best model
print("\nClassification report (best model):")
print(classification_report(y_test, best_pred, target_names=iris.target_names))


Classification report (best model):
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      0.90      0.95        10
   virginica       0.91      1.00      0.95        10

    accuracy                           0.97        30
   macro avg       0.97      0.97      0.97        30
weighted avg       0.97      0.97      0.97        30



In [20]:
print("Confusion matrix (best model):")
print(confusion_matrix(y_test, best_pred))

Confusion matrix (best model):


NameError: name 'confusion_matrix' is not defined

In [21]:
import joblib

In [22]:
# Save best model (SVM)
joblib.dump(models["SVM"], "iris_svm_model.pkl")
joblib.dump(scaler,"iris_scaler.pkl")
print(" Model and scaler saved (iris_svm_model.pkl, iris_scaler.pkl)")


 Model and scaler saved (iris_svm_model.pkl, iris_scaler.pkl)


**🚀 Deployment Options**

Using **Streamlit** (simple web app)

Create app.py:

In [23]:
import streamlit as st
import joblib
import numpy as np

In [24]:
# !pip install streamlit

In [25]:
# !streamlit --version

In [26]:

# Load trained model
model = joblib.load("iris_svm_model.pkl")
scaler = joblib.load("iris_scaler.pkl")

st.title(" Iris Flower Prediction App")
st.write("Enter flower measurements and get the predicted species!")





2025-09-06 21:51:05.218 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.627 
  command:

    streamlit run C:\Users\dell4\anaconda3\Lib\site-packages\ipykernel_launcher.py [ARGUMENTS]
2025-09-06 21:51:05.635 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.635 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.635 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.635 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.635 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


In [27]:
# Input fields
sepal_length = st.number_input("Sepal Length (cm)", min_value=0.0, max_value=10.0, step=0.1)
sepal_width = st.number_input("Sepal Width (cm)", min_value=0.0, max_value=10.0, step=0.1)
petal_length = st.number_input("Petal Length (cm)", min_value=0.0, max_value=10.0, step=0.1)
petal_width = st.number_input("Petal Width (cm)", min_value=0.0, max_value=10.0, step=0.1)


2025-09-06 21:51:05.651 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.651 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.660 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.660 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.660 Session state does not function when running a script without `streamlit run`
2025-09-06 21:51:05.660 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.668 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:05.668 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51

In [28]:
# Predict button
if st.button("Predict"):
    features = np.array([[sepal_length, sepal_width, petal_length, petal_width]])
    features_scaled = scaler.transform(features)
    prediction = model.predict(features_scaled)[0]
    species = ["Setosa ", "Versicolor ", "Virginica "]
    st.success(f"Prediction: **{species[prediction]}**")

2025-09-06 21:51:06.196 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:06.196 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:06.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:06.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:06.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-09-06 21:51:06.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


In [29]:
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0         0  
1         0  
2         0  
3         0  
4         0  
..      ...  
145       2  
146       2  
147       2  
148       2  
149       2  

[150 rows x 5 columns]